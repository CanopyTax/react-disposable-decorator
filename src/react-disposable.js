import React from "react";
import ReactDOM from "react-dom";
import { Observable } from "rx";

const inTestingEnv = typeof jasmine !== "undefined";

export function DisposableDecorator() {
	return function decorator(DecoratedComponent) {
		return class Disposable extends React.Component {
			disposables = [];
			render() {
				return (
					<DecoratedComponent
						{...this.props}
						cancelWhenUnmounted={this.cancelWhenUnmounted}
						cancelAllSubscriptions={this.cancelAllSubscriptions}
					/>
				);
			}
			componentWillUnmount() {
				this.cancelAllSubscriptions();
			}
			cancelWhenUnmounted = cancelWhenUnmounted.bind(this);
			cancelAllSubscriptions = cancelAllSubscriptions.bind(this);
		};
	};
}

export function disposableModal(El, props = {}) {
	let el;

	return Observable.create(observer => {
		el = document.createElement("div");
		document.body.appendChild(el);

		ReactDOM.render(
			<El {...props} close={observer.onCompleted} save={observer.onNext} />,
			el
		);
	}).finally(() => {
		ReactDOM.unmountComponentAtNode(el);
		el.parentNode.removeChild(el);
	});
}

export class DisposableElement extends React.Component {
	disposables = [];
	render() {
		return this.props.children(
			this.cancelWhenUnmounted,
			this.cancelAllSubscriptions
		);
	}
	componentWillUnmount() {
		this.cancelAllSubscriptions();
	}
	cancelWhenUnmounted = cancelWhenUnmounted.bind(this);
	cancelAllSubscriptions = cancelAllSubscriptions.bind(this);
}

function cancelWhenUnmounted(...thingsToCancel) {
	thingsToCancel.forEach(thingToCancel => {
		if (!thingToCancel || typeof thingToCancel.dispose !== "function") {
			throw new Error(
				`cancelWhenUnmounted should be called with one or more disposables (an object with a dispose function)`
			);
		}
		this.disposables.push(thingToCancel);
	});
}

function cancelAllSubscriptions() {
	this.disposables.forEach(disposable => {
		if (disposable && typeof disposable.dispose === "function") {
			disposable.dispose();
		}
	});

	this.disposables = [];
}
