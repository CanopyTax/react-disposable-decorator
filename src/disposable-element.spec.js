import { DisposableElement } from "./react-disposable.js";
import { Observable } from "rx";
import { mount } from "enzyme";
import React from "react";

describe("DisposableElement", function() {
	it("should render", function() {
		function Component({ cancelWhenUnmounted, cancelAllSubscriptions, stuff }) {
			return <div>Hello {stuff}</div>;
		}

		const wrapper = mount(
			<DisposableElement>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</DisposableElement>
		);

		expect(wrapper).toMatchSnapshot();
	});

	it("should cancel subscriptions", function(done) {
		class Component extends React.Component {
			componentDidMount() {
				this.props.cancelWhenUnmounted(
					Observable.timer(100).subscribe(() => {
						fail('should have canceled the subscription');
					})
				);
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		const wrapper = mount(
			<DisposableElement>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</DisposableElement>
		);

		wrapper.unmount();
		setTimeout(done, 200);
	});

	it("should cancel all subscriptions", function(done) {
		class Component extends React.Component {
			componentDidMount() {
				this.props.cancelWhenUnmounted(
					Observable.timer(100).subscribe(() => {
						fail('should have canceled the subscription');
					})
				);

				this.props.cancelWhenUnmounted(
					Observable.timer(50).subscribe(() => {
						fail('should have canceled the subscription');
					})
				);

				this.props.cancelAllSubscriptions();
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		const wrapper = mount(
			<DisposableElement>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</DisposableElement>
		);

		setTimeout(done, 200);
	});

	it("should throw when a non subscription is passed", function() {
		class Component extends React.Component {
			componentDidMount() {
				this.props.cancelWhenUnmounted(
					"this should not work"
				);
			}
			render() {
				return <div>Hello {this.props.stuff}</div>;
			}
		}

		const makeError = () => mount(
			<DisposableElement>
				{(cancelWhenUnmounted, cancelAllSubscriptions) => (
					<Component
						cancelWhenUnmounted={cancelWhenUnmounted}
						cancelAllSubscriptions={cancelAllSubscriptions}
						stuff={1}
					/>
				)}
			</DisposableElement>
		);

		expect(makeError).toThrowError('cancelWhenUnmounted should be called with one or more disposables (an object with a dispose function)');
	});
});
