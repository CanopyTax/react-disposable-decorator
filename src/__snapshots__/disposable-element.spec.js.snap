// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisposableElement should render 1`] = `
<DisposableElement>
  <Component
    cancelAllSubscriptions={[Function]}
    cancelWhenUnmounted={[Function]}
    stuff={1}
  >
    <div>
      Hello 
      1
    </div>
  </Component>
</DisposableElement>
`;
